File: stdlib/kzip/kzip.py
# test642 : stdlib5.kzip

import os
import shutil

import picdt
import ksc

def delpath(path): # delete file/folder if exists
    if os.path.exists(path):
        if os.path.isfile(path):
            os.remove(path)
        else:
            shutil.rmtree(path)

def abspath(tgts): # return stdpath of folder/file list
    temp = [os.path.abspath(x).replace("\\", "/") for x in tgts]
    for i in range( 0, len(temp) ):
        if os.path.isdir( temp[i] ) and temp[i][-1] != "/":
            temp[i] = temp[i] + "/"
    return temp

def getlist(tgts): # get folder/file names from tgts
    lstd, lstf, absf = [ ], [ ], [ ]
    tgts = abspath(tgts)
    count = 0
    for i in tgts:
        if i[-1] == "/":
            if i.count("/") == 1:
                temp = f"LargeVolume{count}/"
                count = count + 1
            else:
                temp = i[0:-1]
                temp = temp[temp.rfind("/") + 1:] + "/"
            getsub(i, temp, lstd, lstf, absf)
        else:
            lstf.append( i[i.rfind("/") + 1:] )
            absf.append(i)
    return lstd, lstf, absf

def getsub(path, prefix, lstd, lstf, absf): # path : stdpath(~/) of root, prefix : ("root/")
    lstd.append(prefix)
    temp = [ x for x in os.listdir(path) ]
    for i in range( 0, len(temp) ):
        if os.path.isdir( path + temp[i] ) and temp[i][-1] != "/":
            temp[i] = temp[i] + "/"
    for i in temp:
        if i[-1] == "/":
            getsub(path + i, prefix + i, lstd, lstf, absf)
        else:
            lstf.append(prefix + i)
            absf.append(path + i)

# zip. tgts : folder/file path, mode ("webp", "png", ""), path ("" -> "./temp570.webp")
def dozip(tgts, mode, path):
    lstd, lstf, absf = getlist(tgts)
    header = bytes("\n".join(lstd), encoding="utf-8")

    writer = ksc.toolbox()
    if mode == "webp":
        writer.prehead = picdt.toolbox().data1
        writer.prehead = writer.prehead + b"\x00" * ( ( 0 - len(writer.prehead) ) % 512 )
    elif mode == "png":
        writer.prehead = picdt.toolbox().data0
        writer.prehead = writer.prehead + b"\x00" * ( ( 0 - len(writer.prehead) ) % 512 )
    else:
        writer.prehead = b""
    writer.subtype = b"kzip"
    writer.reserved = ksc.crc32hash(header) + b"\x00\x00\x00\x00"
    if path == "":
        writer.path = "./temp570.webp"
    else:
        writer.path = path
    delpath(writer.path)

    writer.writef()
    writer.linkf(header)
    for i in range( 0, len(lstf) ):
        writer.linkf( bytes(lstf[i], encoding="utf-8") )
        writer.addf( absf[i] )
    writer.addf("")

# unzip. kzip file path, export folder ("" -> "./temp570/"), check subtype/crc error
def unzip(path, export, chkerr):
    if export == "":
        export = "./temp570/"
    export = export.replace("\\", "/")
    if export[-1] != "/":
        export = export + "/"
    delpath(export)
    os.mkdir(export)

    reader = ksc.toolbox()
    reader.path = path
    reader.predetect = True
    reader.readf()

    with open(path, "rb") as f:
        f.seek(reader.chunkpos[0] + 8)
        header = f.read( reader.chunksize[0] )
        if chkerr:
            if reader.subtype != b"kzip":
                raise Exception("InvalidKZIP")
            if reader.reserved[0:4] != ksc.crc32hash(header):
                raise Exception("InvalidCRC")
            if len(reader.chunkpos) % 2 != 1:
                raise Exception("InvalidChunk")

        if header == b"":
            header = [ ]
        else:
            header = str(header, encoding="utf-8").split("\n")
        for i in header:
            os.mkdir(export + i)

        for i in range(0, (len(reader.chunkpos) - 1) // 2):
            pos = 2 * i + 1
            f.seek(reader.chunkpos[pos] + 8)
            name = str(f.read( reader.chunksize[pos] ), encoding="utf-8")
            size = reader.chunksize[pos + 1]
            f.seek(reader.chunkpos[pos + 1] + 8)
            with open(export + name, "wb") as t:
                for j in range(0, size // 10485760):
                    t.write( f.read(10485760) )
                if size % 10485760 != 0:
                    t.write( f.read(size % 10485760) )

File: stdlib/kzip/kzip.go
// test643 : stdlib5.kzip

package kzip

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"stdlib5/kio"
	"stdlib5/ksc"
	"stdlib5/picdt"
	"strings"
)

// check if path is folder
func isdir(path string) bool {
	temp, err := os.Stat(path)
	if err == nil {
		return temp.IsDir()
	} else {
		return false
	}
}

// get folder/file names from tgts
func getlist(tgts []string) ([]string, []string, []string) {
	for i, r := range tgts {
		temp, _ := filepath.Abs(r)
		temp = strings.Replace(temp, "\\", "/", -1)
		if isdir(temp) && temp[len(temp)-1] != '/' {
			temp = temp + "/"
		}
		tgts[i] = temp
	}

	lstd := make([]string, 0)
	lstf := make([]string, 0)
	absf := make([]string, 0)
	count := 0

	for _, r := range tgts {
		if r[len(r)-1] == '/' {
			temp := ""
			if strings.Count(r, "/") == 1 {
				temp = fmt.Sprintf("LargeVolume%d/", count)
				count = count + 1
			} else {
				temp = r[0 : len(r)-1]
				temp = temp[strings.LastIndex(temp, "/")+1:] + "/"
			}
			getsub(r, temp, &lstd, &lstf, &absf)
		} else {
			lstf = append(lstf, r[strings.LastIndex(r, "/")+1:])
			absf = append(absf, r)
		}
	}
	return lstd, lstf, absf
}

// path : stdpath(~/) of root, prefix : ("root/")
func getsub(path string, prefix string, lstdp *[]string, lstfp *[]string, absfp *[]string) {
	*lstdp = append(*lstdp, prefix)
	fs, _ := os.ReadDir(path)
	temp := make([]string, len(fs))
	for i, r := range fs {
		rr := r.Name()
		if isdir(path+rr) && rr[len(rr)-1] != '/' {
			temp[i] = rr + "/"
		} else {
			temp[i] = rr
		}
	}

	for _, r := range temp {
		if r[len(r)-1] == '/' {
			getsub(path+r, prefix+r, lstdp, lstfp, absfp)
		} else {
			*lstfp = append(*lstfp, prefix+r)
			*absfp = append(*absfp, path+r)
		}
	}
}

// zip. tgts : folder/file path, mode ("webp", "png", ""), path ("" -> "./temp570.webp")
func Dozip(tgts []string, mode string, path string) (ferr error) {
	defer func() {
		errsign := recover()
		if errsign != nil {
			ferr = fmt.Errorf("%s", errsign)
		}
	}()
	lstd, lstf, absf := getlist(tgts)
	header := []byte(strings.Join(lstd, "\n"))

	writer := ksc.Initksc()
	switch mode {
	case "webp":
		writer.Prehead = picdt.Kz5webp()
		writer.Prehead = append(writer.Prehead, make([]byte, (16384-len(writer.Prehead))%512)...)
	case "png":
		writer.Prehead = picdt.Kz5png()
		writer.Prehead = append(writer.Prehead, make([]byte, (16384-len(writer.Prehead))%512)...)
	default:
		writer.Prehead = make([]byte, 0)
	}
	writer.Subtype = []byte("kzip")
	writer.Reserved = append(ksc.Crc32hash(header), 0, 0, 0, 0)
	if path == "" {
		writer.Path = "./temp570.webp"
	} else {
		writer.Path = path
	}
	os.RemoveAll(writer.Path)

	writer.Writef()
	writer.Linkf(header)
	for i, r := range lstf {
		writer.Linkf([]byte(r))
		writer.Addf(absf[i])
	}
	writer.Addf("")
	return nil
}

// unzip. kzip file path, export folder ("" -> "./temp570/"), check subtype/crc error
func Unzip(path string, export string, chkerr bool) error {
	if export == "" {
		export = "./temp570/"
	}
	export = strings.Replace(export, "\\", "/", -1)
	if export[len(export)-1] != '/' {
		export = export + "/"
	}
	var err error
	err = os.RemoveAll(export)
	if err != nil {
		return err
	}
	err = os.Mkdir(export, os.ModePerm)
	if err != nil {
		return err
	}

	reader := ksc.Initksc()
	reader.Path = path
	reader.Predetect = true
	err = reader.Readf()
	if err != nil {
		return err
	}

	f, err := kio.Open(path, "r")
	if err == nil {
		defer f.Close()
	} else {
		return err
	}
	f.Seek(int64(reader.Chunkpos[0]+8), 0)
	header, _ := kio.Read(f, reader.Chunksize[0])
	if chkerr {
		if !kio.Bequal(reader.Subtype, []byte("kzip")) {
			return errors.New("InvalidKZIP")
		}
		if !kio.Bequal(reader.Reserved[0:4], ksc.Crc32hash(header)) {
			return errors.New("InvalidCRC")
		}
		if len(reader.Chunkpos)%2 != 1 {
			return errors.New("InvalidChunk")
		}
	}

	var lstd []string
	if len(header) == 0 {
		lstd = make([]string, 0)
	} else {
		lstd = strings.Split(string(header), "\n")
	}
	for _, r := range lstd {
		err = os.Mkdir(export+r, os.ModePerm)
		if err != nil {
			return err
		}
	}

	for i := 0; i < (len(reader.Chunkpos)-1)/2; i++ {
		pos := 2*i + 1
		f.Seek(int64(reader.Chunkpos[pos]+8), 0)
		nmb, _ := kio.Read(f, reader.Chunksize[pos])
		name := string(nmb)
		size := reader.Chunksize[pos+1]
		f.Seek(int64(reader.Chunkpos[pos+1]+8), 0)

		t, err := kio.Open(export+name, "w")
		if err != nil {
			return err
		}
		for j := 0; j < size/10485760; j++ {
			temp, _ := kio.Read(f, 10485760)
			kio.Write(t, temp)
		}
		if size%10485760 != 0 {
			temp, _ := kio.Read(f, size%10485760)
			kio.Write(t, temp)
		}
		t.Close()
	}
	return nil
}
